src/app/utils/dataLoaderUtil.js

```diff
@@ -0,0 +1,127 @@
+import Actions from '@Actions';
+import { ajaxCall } from '@utils';
+import appConfig from '@appConfig';
+import Store from '@Store';
+
+const pathInstructions = [
+  {
+    expression: /\/research\/collections\/shared-collection-catalog\/bib\/([cp]?b\d*)/,
+    pathType: 'bib',
+  },
+  {
+    expression: /\/research\/collections\/shared-collection-catalog\/search\?(.*)/,
+    pathType: 'search',
+  },
+  {
+    expression: /\/research\/collections\/shared-collection-catalog\/hold\/request\/([^/]*)/,
+    pathType: 'holdRequest',
+  },
+];
+
+const routePaths = {
+  bib: `${appConfig.baseUrl}/api/bib`,
+  search: `${appConfig.baseUrl}/api`,
+  holdRequest: `${appConfig.baseUrl}/api/hold/request/:bibId-:itemId`,
+};
+
+const routesGenerator = location => ({
+  bib: {
+    apiRoute: (matchData, route) => `${route}?bibId=${matchData[1]}`,
+    actions: [Actions.updateBib],
+    errorMessage: 'Error attempting to make an ajax request to fetch a bib record from ResultsList',
+  },
+  search: {
+    apiRoute: (matchData, route) => `${route}?${matchData[1]}`,
+    actions: [
+      data => Actions.updateSearchResults(data.searchResults),
+      () => Actions.updatePage(location.query.page || 1),
+      () => Actions.updateSearchKeywords(location.query.q),
+      data => Actions.updateFilters(data.filters),
+      () => {
+        const {
+          sort,
+          sort_direction,
+        } = location.query;
+        Actions.updateSortBy(`${sort}_${sort_direction}`);
+      },
+    ],
+    errorMessage: 'Error attempting to make an ajax request to search',
+  },
+  holdRequest: {
+    apiRoute: (matchData, route) => route.replace(':bibId-itemId', matchData[1]),
+    actions: [
+      data => Actions.updateBib(data.bib),
+      data => Actions.updateDeliveryLocations(data.deliveryLocations),
+      data => Actions.updateIsEddRequestable(data.isEddRequestable),
+    ],
+    errorMessage: 'Error attempting to make ajax request for hold request',
+  },
+});
+
+const reducePathExpressions = location => (acc, instruction) => {
+  const {
+    pathname,
+    search,
+  } = location;
+
+  const {
+    matchData,
+    pathType,
+  } = acc;
+
+  const newMatchData = (pathname + search).match(instruction.expression);
+  return {
+    matchData: matchData || newMatchData,
+    pathType: pathType || (newMatchData ? instruction.pathType : null),
+  };
+};
+
+function loadDataForRoutes(location, req, RouteMethods) {
+  const routes = routesGenerator(location);
+  const {
+    matchData,
+    pathType,
+  } = pathInstructions
+    .reduce(reducePathExpressions(location), { matchData: null, pathType: null });
```

Comment:
It doesn't seem like `reduce` is the right function to use. Nothing is being accumulated and returned but rather you're looping through all the path instructions until the location matches the expression regex and returning that. So in `reducePathExpressions`, if the location is `/bib`, then you'll get the `matchData` and `pathType` right away but keep looking through the rest of path instructions (and since nothing will match, then `matchData` and `pathType` never change). It seems like it might be better to use `find` instead of going through the whole array once  you find the location that matches.

Reply:
Refactored this to use find